python/lsst/ip/isr/deferredCharge.py

```diff
@@ -0,0 +1,639 @@
+
+    def run(self, exposure, overscans, ctiCalib, gains=None):
+        """Correct deferred charge/CTI issues.
+
+        Parameters
+        ----------
+        exposure : `lsst.afw.image.Exposure`
+            Exposure to correct the deferred charge on.
+        gains : `dict` [`str`, `float`]
+            A dictionary, keyed by amplifier name, of the gains to
+            use.  If gains is None, the nominal gains in the amplifier
+            object are used
+        ctiCalib : `lsst.ip.isr.DeferredChargeCalib`
+            Calibration object containing the charge transfer
+            inefficiency model.
+
+        Returns
+        -------
+        exposure : `lsst.afw.image.Exposure`
+            The corrected exposure.
+        """
+        image = exposure.getMaskedImage().getImage()
+        detector = exposure.getDetector()
+        if self.config.useGains:
+            if gains is None:
+                gains = {amp.getName(): amp.getGain() for amp in detector.getAmplifiers()}
+        else:
+            gains = {amp.getName(): 1.0 for amp in detector.getAmplifiers()}
+
+        with gainContext(exposure, image, True, gains):
+            for amp in detector.getAmplifiers():
+                ampName = amp.getName()
+
+                ampImage = image[amp.getRawBBox()]
+                if self.config.zeroUnusedPixels:
+                    # We don't apply overscan subtraction, so zero these
+                    # out for now.
+                    ampImage[amp.getRawParallelOverscanBBox()].getArray()[:, :] = 0.0
+                    ampImage[amp.getRawSerialPrescanBBox()].getArray()[:, :] = 0.0
+
+                # The algorithm expects that the readout corner is in
+                # the lower left corner.  Flip it to be so:
+
+                ampData = self.flipData(ampImage.getArray(), amp)
+
+                if ctiCalib.driftScale[ampName] > 0.0:
+                    correctedAmpData = self.local_offset_inverse(ampData,
+                                                                 ctiCalib.driftScale[ampName],
+                                                                 ctiCalib.decayTime[ampName],
+                                                                 self.config.nPixelOffsetCorrection)
+                else:
+                    correctedAmpData = ampData.copy()
+
+                correctedAmpData = self.local_trap_inverse(correctedAmpData,
+                                                           ctiCalib.serialTraps[ampName],
+                                                           ctiCalib.globalCti[ampName],
+                                                           self.config.nPixelTrapCorrection)
+
+                # Undo flips here.  The method is symmetric.
+                correctedAmpData = self.flipData(correctedAmpData, amp)
+                image[amp.getBBox()].getArray()[:, :] = correctedAmpData[:, :]
+
+        return exposure
+
+    @staticmethod
+    def flipData(ampData, amp):
```

Comment:
github is convinced that I must comment here (and keeps warping my browser to this box).  I don't think there's any problem here.  It's in numpy simply because the downstream code expects a numpy ndarray as input, so it was easier to work in that space.